Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -265,114 +234,132 @@ final class ConnectionPool {
-        ///     - parentPool: The pool this provider belongs to
-        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, parentPool: ConnectionPool) {
+        ///     - pool: The pool this provider belongs to
+        init(key: ConnectionPool.Key, eventLoop: EventLoop, configuration: HTTPClient.Configuration, pool: ConnectionPool) {
             self.eventLoop = eventLoop
             self.configuration = configuration
             self.key = key
-            self.parentPool = parentPool
-            self.state = State(eventLoop: eventLoop, parentPool: parentPool, key: key)
+            self.pool = pool
         }
 
         deinit {
-            assert(self.state.activity == .closed, "Non closed on deinit")
-            assert(self.state.availableConnections.isEmpty, "Available connections should be empty before deinit")
-            assert(self.state.leased == 0, "All leased connections should have been returned before deinit")
-            assert(self.state.waiters.count == 0, "Waiters on deinit: \(self.state.waiters)")
+            // TODO:
+//            assert(self.state.activity == .closed, "Non closed on deinit")
+//            assert(self.state.availableConnections.isEmpty, "Available connections should be empty before deinit")
+//            assert(self.state.leased == 0, "All leased connections should have been returned before deinit")
+//            assert(self.state.waiters.count == 0, "Waiters on deinit: \(self.state.waiters)")
         }
 
         var description: String {
-            return "HTTP1ConnectionProvider { key: \(self.key), state: \(self.state) }"
+            return "HTTP1ConnectionProvider { key: \(self.key) }"
         }
 
         func getConnection(preference: HTTPClient.EventLoopPreference) -> EventLoopFuture<Connection> {
-            self.activityPrecondition(expected: [.opened])
-            let action = self.stateLock.withLock { self.state.connectionAction(for: preference) }
-            switch action {
-            case .leaseConnection(let connection):
-                return connection.removeIdleConnectionHandlersForLease().flatMapError { _ in
-                    connection.closeFuture.flatMap { // We ensure close actions are run first
-                        let defaultEventLoop = self.stateLock.withLock {
-                            self.state.defaultEventLoop
+            let promise = self.eventLoop.makePromise(of: Connection.self)
+
+            self.lock.withLockVoid {
+                if let connection = self.availableConnections.popFirst() {
+                    connection.isLeased = true
+                    // check if we can vend this connection to caller
+                    connection.cancelIdleTimeout().flatMapError { error in
+                        if error is Connection.InactiveChannelError {
+                            return self.makeConnection(on: preference.bestEventLoop ?? self.eventLoop)
```

Comment:
are we _sure_ this doesn't call out immediately? If anything in there succeeds or fails a future, this isn't valid as we'd call out with the lock held.

Reply:
Same question applies to _all_ other calls we make with the lock held. I think the idea of the `struct State` was to have a place where all the computation goes (under a lock). The computation would then return what the next correct step is that needs to be performed (not under a lock).

Reply:
I did hit that (calling out and then deadlocking, see line 328), so yes, I probably will revert some of that state/action stuff...

Reply:
Yeah, I know it's annoying but it's the only way to retain sanity: Separate the state from the actions, and then it'll be cool.

A good way to spot bugs is: If an action is returned but the state is _not_ changed, it's probably a bug. Because someone else could then race in and get the same action before the first action has run (and advanced the state with requiring the lock). By the time the second action is then executing the state has then changed under its feet and problems happen.